crat.1
.TH crat 1 "1.0" "Internet Connection, Inc." "Ad-Hoc Clusters"
.SH NAME
crat \- log ratout alerts from dissatisfied cexec runs
.SH SYNOPSIS
.B crat keys.pub...

Logs errors that cexec detects. This tool should be used on every
cluster to detect failed jobs and failing hosts.
.SH DESCRIPTION
.B crat
writes log entries to standard ouput. They should be saved by a
tool like
.I multilog
or
.I splogger
but they are not covered here.

Log entries look like this:

.I 10.0.0.5 rats out 10.0.0.3 for foo.pub
 
.I 10.0.0.6 completed by 10.0.0.3 for foo.pub exit code 0
 
.I 10.0.0.6 completed by 10.0.0.2 for foo.pub exit code 1

It is left as an exercise to the imagination of the reader what
to do with this information.

.B NOTE
that failures are normal. You should not be worried seeing one
or two "rats out" lines periodically. You should be worried
about several dozen lines worth, however.

.SH ENVIRONMENT
.B $GROUP
\- contains the "group name" - an IP address (possibly multicast) representing
the address to be used for receiving alerts.
.I Default: 255.255.255.255
.SH EXIT CODES
.B crat
normally doesn't exit. All exit codes are errors.
.SH SEE ALSO
ckeygen(1), cexec(1), cservice(1), cproxy(1)


CHANGELOG
2008-08-07 18:02  geocar

	* Makefile, bio.c, build-cross.sh, cexec.c, ckeygen.c, cproxy.c,
	  crat.c, cservice.c, deploy.sh, hash.c, keyfile.c, load.c,
	  multipipe.c, netnum.c, netnum.h, parity.c, sha1.c, sha1.h,
	  strline.c, strline.h, types.h: port to 64bit system

2007-03-02 22:05  geocar

	* deploy.sh: amd64

2007-01-30 21:27  geocar

	* cexec.c: wee

2006-12-08 22:48  geocar

	* bio.c, cexec.c, cservice.c, deploy.sh: moving it around...

2006-12-08 17:48  geocar

	* bio.c, build-cross.sh, cexec.c, cio.c, cservice.c, deploy.sh: fix
	  base when doing copyout; dont close on write (sigpipe helps
	  noone) and sparc is dead. oh, and nul-group is all-group

2006-07-31 06:46  geocar

	* cservice.c: support $BOOST mode

2006-07-31 06:46  geocar

	* load.c: use gnu load

2006-06-23 06:59  geocar

	* cexec.c: dont regenerate parity (we cannot remember long enough)

2006-06-23 06:59  geocar

	* cservice.c: fix delays, and make pubkey toplevel so we can test

2006-06-23 05:27  geocar

	* Makefile: uses sockop

2006-06-23 05:27  geocar

	* cexec.c: add testing feature, and skipto to try next-best host

2006-06-23 05:26  geocar

	* cservice.c: check signature for safety and use sockop_nonblock

2006-06-23 05:26  geocar

	* hash.c: zero s1/s2 barrels

2006-06-23 05:26  geocar

	* keyfile.c: check zero length vlpoint

2006-06-23 05:26  geocar

	* cio.c, sockop.c: bleh, convenience i suppose

2005-11-08 07:51  geocar

	* cexec.c: remove did_io optimization; it's complicated

2005-03-17 10:07  geocar

	* ckeygen.1: adding documentation about -stdin flag

2005-03-17 10:06  geocar

	* ckeygen.c: generate keys based on stdin with -stdin option now...

2005-03-17 10:06  geocar

	* cexec.c: did_io when writing too...

2005-03-01 12:58  geocar

	* cservice.c: memalloc isn't fatal PEER_vars, don't attempt to dup2
	  a -1 fd (bitmap)

2005-03-01 11:31  geocar

	* cproxy.c: don't waste file descriptors all wily-nilly, also
	  multihomed SO_BINDTODEVICE capable boxes are now assisted...

2005-03-01 11:31  geocar

	* cexec.c: put lseek hack back in

2005-02-17 08:11  geocar

	* parity.c: use urandom instead of random; the lifetime of the
	  check-bits are so short that other attacks are more likely to be
	  useful

2005-02-17 07:59  geocar

	* cservice.c: do bit flipping if near or far is running on a really
	  strange platform

2005-02-17 07:58  geocar

	* cexec.c: return didio hack

2005-02-17 07:58  geocar

	* cexec.c, cluster.h: move bytesex constant out of cexec

2005-02-17 07:57  geocar

	* bio.c: sink output in blast and copy

2005-02-17 06:27  geocar

	* LICENSE, Makefile, cexec.c, cio.c, cio.h, cservice.c, keyfile.c:
	  moved cread/cwrite out of keyfile and changed write-hack to use
	  4bytes instead of a vlpoint. the write-hack might prove useful
	  someday...

2005-02-16 14:40  geocar

	* cexec.c, cservice.c: turn keepalive back on

2005-02-16 13:02  geocar

	* cexec.c, cservice.c: don't use keepalive anymore, protect
	  main-channel for sanity, and add an extra vlpoint-exchange just
	  to force data both ways on the socket before working/forking

2005-02-16 13:00  geocar

	* bio.c: protect fcntl changes

2005-02-16 12:50  geocar

	* bio.c: turn off nonblock when blasting, and honor EAGAIN on
	  bio_more

2005-02-16 12:47  geocar

	* keyfile.c: turn off nonblock when in cread/cwrite

2005-02-16 09:54  geocar

	* cservice.1, cservice.c: set $REMOTE_ADDR, $REMOTE_PORT,
	  $SERVER_ADDR, $SERVER_PORT so cservice child can do things like
	  ident-check the cexec peer if it likes

2005-02-16 09:53  geocar

	* crat.c: unknown keys are listed in hex for brevity

2005-02-16 09:53  geocar

	* strline.c, strline.h: support for hex numbers too

2005-02-16 08:29  geocar

	* LICENSE, Makefile, cexec.c, crat.c, cservice.c, fdm.c, pair.c,
	  sockop.c, sockop.h: move sockops into a separate file

2005-02-16 08:28  geocar

	* cproxy.c: bugfix; detect hangup on write; move expires operation
	  before gathering fds so cproxy doesn't bomb

2005-02-15 10:44  geocar

	* cexec.c: restart processing if we haven't done any I/O but still
	  got an error,

2005-02-15 10:44  geocar

	* bio.c: treat EINTR and EAGAIN the same

2005-02-15 10:43  geocar

	* cservice.c: map fd masks correctly; increased timeouts, don't try
	  and read from wronly fds, use PIPE_BUF for buffers, and send
	  close alerts for write-errors

2005-02-15 10:42  geocar

	* fdm.c: map rdonly file descriptors correctly now; treat ttys
	  specially

2005-02-15 10:41  geocar

	* keyfile.c: treat EINTR and EAGAIN the same; also be pendantic
	  about short != 16 bits

2005-02-14 17:09  geocar

	* cexec.c: use a different (better?) announcement schedule

2005-02-14 17:09  geocar

	* cproxy.c: un-nonblock fds, turn on keepalive/ndelay on both
	  sides, honor EAGAIN

2005-02-14 17:09  geocar

	* cservice.1, cservice.c: lower default concurrency, drop
	  unnecessary checks, and recover from autorat failures faster

2005-02-14 16:35  geocar

	* load.c: system_load() isn't required to actually function

2005-02-14 11:01  geocar

	* cexec.1, ckeygen.1, cproxy.1, crat.1, cservice.1: adding
	  documentation

2005-02-14 08:57  geocar

	* cexec.c: don't waste an fd

2005-02-14 08:47  geocar

	* cexec.c, cservice.c: added pingpong detection, configurable
	  timeouts, configurable port ranges, using simplexes instead of
	  duplexes where possible, sending fd usage map, and debugging
	  nonsense for cexec

2005-02-14 08:43  geocar

	* crat.c, strline.c, strline.h: move string opers stuff

2005-02-14 08:42  geocar

	* fdm.c: fixes in detecting file descriptors on Linux

2005-02-14 08:32  geocar

	* pair.c: turn in so_linger on our pair links.

2005-02-14 08:30  geocar

	* parity.c: use /dev/random if available

2004-12-16 07:21  geocar

	* cservice.c: moving waitpid chunk earlier,change autorat timeouts,
	  and disable need_delay after delay (procrastinating autorats)

2004-12-16 06:37  geocar

	* cservice.c: default maxchildren now 100

2004-12-16 06:31  geocar


crat.c
#include <sys/types.h>
#include <sys/time.h>
#include <sys/select.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <fcntl.h>

#include "cluster.h"
#include "netnum.h"

#include "ec_crypt.h"
#include "ec_vlong.h"

#include "strline.h"

void usage(void)
{
	int i,r;
#define M "Usage: crat pubkey... > log\n"
	for (i = 0; i < sizeof(M)-1;) {
		do {
			r = write(2, M+i,(sizeof(M)-1)-i);
		} while (r == -1 && errno == EINTR);
		if (r < 1) exit(255);
		i += r;
	}
	exit(0);
#undef M
}
int main(int argc, char *argv[])
{
	struct ip_mreq gamr;
	unsigned char buf[512];
	unsigned char *outline;
	unsigned char rattest[9];
	unsigned int hashbuf_size;
	unsigned char *hashbuf;
	struct sockaddr_in sin;
	unsigned char *hashes;
	unsigned int hv;
	vl_point pub;
	unsigned int outline_len, linelen;
	int i, r, fd;
	unsigned char *z;
	in_addr_t ina;
	char *q;

	if (argc < 1) usage();
	hashes = (unsigned char *)malloc(8 * (argc-1));
	if (!hashes)exit(111);
	for (i = 1, outline_len = 0; i < argc; i++) {
		r = strlen(argv[i]);
		if (r > outline_len) outline_len = r;
		fd = open(argv[i], O_RDONLY);
		if (fd == -1) {
			q = (char*)malloc(strlen(argv[i])+5);
			if (!q) exit(111);
			strcpy(q, argv[i]);
			strcat(q, ".pub");
			fd = open(q, O_RDONLY);
			if (fd == -1) usage();
			(void)free(q);
		}
		if (!vlpoint_read(pub, fd)) exit(111);
		(void)close(fd);
		
		hashbuf_size = vlpoint_size(pub);
		hashbuf = (char *)malloc(hashbuf_size);
		if (!hashbuf) exit(111);
		vlpoint_pack(pub, hashbuf, hashbuf_size);
		hv = hash(hashbuf, hashbuf_size);
		(void)free(hashbuf);
		put_uint32(hashes+((i-1)*4), hv);
	}
	outline_len += 15; /* strings */
	outline_len += 32; /* two ip addresses */
	outline_len += 128;/* some buffer space in case some bonehead isn't
			      paying attention... :) */
	outline = (char *)malloc(outline_len);
	if (!outline) exit(111);

	fd = socket(AF_INET, SOCK_DGRAM, 0);
	if (fd == -1) exit(111);

	memset(&sin, 0, sizeof(sin));
	q = (char *)getenv("GROUP");
	if (!q) q = "255.255.255.255";
	ina = inet_addr(q);
	memcpy(&sin.sin_addr, &ina, sizeof(in_addr_t));

	i = *((unsigned char *)&ina);
	if (i >= 224 && i < 255) {
		/* multicast */
		memset(&gamr, 0, sizeof(gamr));
		memcpy(&gamr.imr_multiaddr, &ina, sizeof(in_addr_t));
		i = 1; (void)setsockopt(fd, IPPROTO_IP, IP_MULTICAST_LOOP, &i, sizeof(i));
		i = 31; (void)setsockopt(fd, IPPROTO_IP, IP_MULTICAST_TTL, &i, sizeof(i));
		/* we might already have joined... */
		i=1;
		(void)setsockopt(fd, IPPROTO_IP, IP_ADD_MEMBERSHIP,
				 &gamr,sizeof(gamr));
	}
	sockop_broadcast(fd, 1);
	sockop_reuseaddr(fd, 1);
	sin.sin_family = AF_INET;
	sin.sin_port = htons(CLUSTER_PORT);
	if (bind(fd, (struct sockaddr *)&sin, sizeof(sin)) == -1) exit(111);

	for (;;) {
		i = sizeof(struct sockaddr_in);
		r = recvfrom(fd, buf, sizeof(buf), 0,
				(struct sockaddr *)&sin, &i);
		if (r != 9 && r != 8) continue; /* ignore this message */
		memcpy(rattest, buf, 4);
		q = 0;
		for (i = 1; i < argc; i++) {
			memcpy(rattest+4, hashes+((i-1)*4), 4);
			put_uint32(rattest+4, hash(rattest, 8));
			if (memcmp(rattest, buf, 8) == 0) {
				q = argv[i];
				break;
			}
		}

		z = outline;
		z += ip_print(z, sin.sin_addr.s_addr);
		if (r == 8) {
			z += static_str(z, " rats out ");
		} else {
			z += static_str(z, " completed by ");
		}
		memcpy(&hv, rattest, 4);
		z += ip_print(z, hv);
		z += static_str(z, " for ");
		if (q) {
			z += static_str(z, q);
		} else {
			z += static_str(z, "unknown key (0x");
			memcpy(&hv, buf+4, 4);
			z += hexnumber_print(z, hv);
			z += static_str(z, ")");
		}
		if (r == 9) {
			z += static_str(z, " exit code ");
			z += number_print(z, buf[8]);
		}
		z += static_str(z, "\n");

		linelen = z - outline;
		for (i = 0; i < linelen;) {
			do {
				r = write(1, outline+i, linelen-i);
			} while (r == -1 && errno == EINTR);
			if (r < 1) exit(255);
			i += r;
		}
	}
}

